File: letthem/chap8/1932.py
n = int(input())
dp = []

for i in range(n):
  dp.append(list(map(int,input().split())))

for i in range(1,n):
  for j in range(0, i+1):
    if j == 0: 
      dp[i][0] += dp[i-1][0] # 0열끼리 더하기
    elif j == i: 
      dp[i][j] += dp[i-1][j-1] # i열끼리 더하기
    else: # 그 외
      dp[i][j] += max(dp[i-1][j-1], dp[i-1][j]) # 위에서 오는 수 중 큰 수

print(max(dp[n-1])) # 가장 마지막 행(다 더해진 행)에서 가장 큰 수
    

File: chae-dahee/ch08/1932.js
const input = require("fs")
  .readFileSync("chae-dahee/ch08/1932.txt")
  .toString()
  .trim()
  .split("\n");
const n = parseInt(input[0]); // 삼각형 높이

const p = new Array(n).fill().map(() => []); // 입력 삼각형 값들과 dp 배열로 사용하기 위한 2차원 배열

// 피라미드 만들기
for (let i = 1; i <= n; i++) {
  const curRow = input[i].split(" ").map(Number);
  p[i - 1].push(...curRow);
}

for (let i = 1; i < n; i++) {
  // 가로 변
  for (let j = 0; j < p[i].length; j++) {
    // 단위 개수
    if (j === 0) {
      // 제일 왼쪽
      p[i][j] = p[i - 1][j] + p[i][j];
    } else if (j === p[i].length - 1) {
      // 제일 오른쪽
      p[i][j] = p[i - 1][j - 1] + p[i][j];
    } // 그 외
    else p[i][j] = Math.max(p[i - 1][j - 1], p[i - 1][j]) + p[i][j];
  }
}

console.log(Math.max(...p[n - 1]));
